### DISCUSSION TITLE
Convert the solution in Function class and assembled operator to numpy array or scipy.csr_matrix vice and versa

### DISCUSSION THREAD

**Discussion by JunmingDuan** [firedrakeproject/firedrake#3103] [Firedrake support] [answered]
Dear community,

I use firedrake to solve a problem: I can build the operator A and RHS b, and solve A*x=b to get x, which is Function.
Now I need to get the matrix and vector of A and x in scipy.csr_matrix and numpy array form, and also backward.
Could you help me on this?

Thank you in advance!
Junming

**Comment by connorjward** [comment, marked as answer]
## Matrix

I have been hacking around and found the following solution to get a scipy `csr_matrix` from a Firedrake `Matrix`:

```py
petsc_mat = firedrake_matrix.M.handle
ai, aj, av = petsc_mat.getValuesCSR()
scipy_matrix = scipy.sparse.csr_matrix((av, aj, ai))
```

Note that this should probably not be attempted in parallel.

For the inverse operation see [this post](https://stackoverflow.com/questions/15442632/scipy-sparse-matrices-as-an-input-for-petsc4py).

## Vector

Firedrake `Function`s are natively stored as numpy arrays so this is a simple operation. You just want `function.dat.data`
